### PR TITLE
Disposing Asynchronously

### PR DESCRIPTION
**Pull request by sburmanoctopus**
[sc-53211]

# Background
As part of making Halibut async, we need to ensure we dispose of things asynchronously as well.

# Results

Related to https://github.com/OctopusDeploy/Issues/issues/8266

## Before

The [following places](https://docs.google.com/spreadsheets/d/1sxtq-K2c8h2Z0ZUFvlouyNTgfPJT0iQzasOJlcsbJEI/edit#gid=0) were found to have paths that would require async disposal. These were being disposed synchronously.

## After
The green classes in the spreadsheet are not disposed asynchronously.

We did not change `using` statements to `await using' statements in this PR. This touched over 20 files, so will be assessed and implemented in a different PR.

To ensure .NET 4.8 code could compile and use `IAsyncDisposable`, we had to import the `Microsoft.Bcl.AsyncInterfaces` package.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut/Transport/Streams/AsyncDisposableStream.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Halibut.Transport.Streams
+{
+    // Streams do not implement IAsyncDisposable in .NET 4.8. So to prevent putting #if everywhere, deal with all this here.
+    public abstract class AsyncDisposableStream : Stream
+#if NETFRAMEWORK
+        , IAsyncDisposable
+#endif
```
Comment (acodrington) [inline source/Halibut/Transport/Streams/AsyncDisposableStream.cs:11]: Truly, truly awful 😆 

It's good we're keeping this hidden

**Review by acodrington — APPROVED** [review]
Yeah looks good dude, nice one!

File: source/Halibut/IHalibutRuntime.cs [hunk 2]
```diff
@@ -15,7 +15,7 @@ public enum UnauthorizedClientConnectResponse
         TrustAndAllowConnection
     }
 
-    public interface IHalibutRuntime : IDisposable
+    public interface IHalibutRuntime : IDisposable, IAsyncDisposable
```
Comment (nathanwoctopusdeploy) [inline source/Halibut/IHalibutRuntime.cs:18]: Are we missing some test coverage/some of or tests using the DisposeAsync method?

E.g. `LatestClientAndLatestServiceBuilder.ClientAndService` calls Dispose on the HalibutRuntimes. We could cover both paths by making it Dispose when we are sync and DisposeAsync when we are forcing it to be async

File: source/Halibut/Transport/ConnectionManager.cs [hunk 3]
```diff
@@ -143,6 +143,10 @@ public void Dispose()
             IsDisposed = true;
         }
 
+        public ValueTask DisposeAsync()
+        {
+            throw new NotImplementedException("Should not be called when async Halibut is not being used.");
```
Comment (nathanwoctopusdeploy) [inline source/Halibut/Transport/ConnectionManager.cs:148]: Nit: I haven't got a strong preference, but we have 2 different ways of saying don't call this cuz your sync/async via `NotSupportedException` and then `NotImplementedException` types.

Should we keep the exception type consistent?

image

Comment (sburmanoctopus) [reply source/Halibut/Transport/ConnectionManager.cs:148]: We totally should! Done 🙂

File: source/Halibut/Transport/ConnectionManagerAsync.cs [hunk 4]
```diff
@@ -99,7 +119,7 @@ public async Task<IConnection> AcquireConnectionAsync(ExchangeProtocolBuilder ex
         Tuple<IConnection, Action> CreateNewConnection(ExchangeProtocolBuilder exchangeProtocolBuilder, IConnectionFactory connectionFactory, ServiceEndPoint serviceEndpoint, ILog log, CancellationToken cancellationToken)
         {
             var lazyConnection = new Lazy<IConnection>(() => connectionFactory.EstablishNewConnection(exchangeProtocolBuilder, serviceEndpoint, log, cancellationToken));
-            var connection = new DisposableNotifierConnection(lazyConnection, OnConnectionDisposed);
+            var connection = new DisposableNotifierConnection(lazyConnection, OnConnectionDisposed, OnConnectionDisposedAsync);
```
Comment (nathanwoctopusdeploy) [inline source/Halibut/Transport/ConnectionManagerAsync.cs:122]: Note: This sync implementation is slated for removal for the context of other readers.

File: source/Halibut/Transport/Observability/ByteCountingStream.cs [hunk 5]
```diff
@@ -45,6 +46,14 @@ protected override void Dispose(bool disposing)
             }
         }
 
+        public override async ValueTask DisposeAsync()
```
Comment (nathanwoctopusdeploy) [inline source/Halibut/Transport/Observability/ByteCountingStream.cs:49]: The Dispose and DisposeAsync differ in their implementation a bit. Should they be consistent?

**Review by nathanwoctopusdeploy — APPROVED** [review]
LGTM 👍

File: source/Halibut/IHalibutRuntime.cs (hunk 2 above)
Comment (sburmanoctopus) [reply source/Halibut/IHalibutRuntime.cs:18]: Good thinking. Done!

**Pull request merged**